bitwarden_access_manager.py

```diff
@@ -0,0 +1,737 @@
+                self._http_get(f"/public/collections/{collection_id}")
+            )
+
+            group_accesses_data = tuple(
+                sorted(self.get_collection_groups(collection_data["groups"]))
+            )
+
+            if len(group_accesses_data) > 0:
+                group_accesses = group_accesses_data
+
+            if collection_id in collections_members:
+                member_accesses = tuple(sorted(collections_members[collection_id]))
+
+            yield Collection(
+                id=collection["id"],
+                external_id=collection["externalId"],
+                member_access=member_accesses,
+                group_access=group_accesses,
+            )
+
+    def get_group_members(self, id: str, name: str) -> Iterable[GroupMember]:
+        members = json.load(self._http_get(f"/public/groups/{id}/member-ids"))
+
+        for member in members:
+            member = json.load(self._http_get(f"/public/members/{member}"))
+            yield GroupMember(
+                member_id=member["id"],
+                member_name=member["name"],
+                group_name=name,
+            )
+
+    def set_member_type(self, type_id: int) -> MemberType:
+        int_to_member_type: Dict[int, MemberType] = {
+            0: MemberType.OWNER,
+            1: MemberType.ADMIN,
+            2: MemberType.USER,
+            3: MemberType.MANAGER,
+            4: MemberType.CUSTOM,
+        }
+        return MemberType(int_to_member_type[type_id])
+
+    def get_members(
+        self,
+    ) -> tuple[List[Member], Dict[str, List[MemberCollectionAccess]]]:
+        data = self._http_get(f"/public/members")
+        members = json.load(data)
+
+        members_result: List[Member] = []
+        collection_access: Dict[str, List[MemberCollectionAccess]] = {}
+
+        for member in members["data"]:
+            type = self.set_member_type(member["type"])
+            m = Member(
+                id=member["id"],
+                name=member["name"],
+                email=member["email"],
+                type=type,
+                access_all=member["accessAll"],
+            )
+            members_result.append(m)
+
+            collections = json.load(self._http_get(f"/public/members/{member['id']}"))[
+                "collections"
+            ]
+            if type != MemberType.OWNER and type != MemberType.ADMIN:
+                for collection in collections:
+                    access = self.check_access(collection["readOnly"])
+
+                    if collection["id"] not in collection_access:
+                        collection_access[collection["id"]] = [
+                            MemberCollectionAccess(name=member["name"], access=access)
+                        ]
+                    else:
+                        collection_access[collection["id"]].append(
+                            MemberCollectionAccess(name=member["name"], access=access)
+                        )
```

Comment:
```suggestion
                    collection_access[collection["id"]].append(
                        MemberCollectionAccess(name=member["name"], access=access)
                    )
```
This is a very common pattern, append something to a list at a given key, and create the list if it does not exist. With `defaultdict` this becomes a bit simpler. In the constructor we provide a default factory, in this case `lambda: []`, a function that returns an empty list. And then if you access any key, if it exists it returns the value like a regular dict, but if the key does not exist, it doesn’t raise a `KeyError` but it calls the default factory and inserts it at that key, so you can use it without ever having to worry about the `KeyError`.

Reply:
https://github.com/ChorusOne/github-access-manager/pull/10/commits/cc2c08119cffefd8a6315fcd0fef9dcd40e5dcd0